fawltydeps/packages.py

```diff
@@ -114,36 +114,71 @@ def lookup_packages(self, package_names: Set[str]) -> Dict[str, Package]:
+    ) -> None:
         self.mapping_path = mapping_path
-        if not self.mapping_path.is_file():
+        if self.mapping_path and not self.mapping_path.is_file():
             raise UnparseablePathException(
                 ctx="Given mapping path is not a file.", path=self.mapping_path
             )
+        self.custom_mapping = custom_mapping
         # We enumerate packages declared in the mapping _once_ and cache the result here:
         self._packages: Optional[Dict[str, Package]] = None
 
     @property
     @calculated_once
     def packages(self) -> Dict[str, Package]:
-        """Gather mapping from a mapping file in a given path.
+        """Gather a custom mapping given by a user.
+
+        Mapping may come from two sources:
+        * _mapping_path_ which is a file in a given path, which is parsed to a ditionary
+        * _custom_mapping_ which is a dictionary of package to imports mapping.
 
         This enumerates the available packages  _once_, and caches the result for
         the remainder of this object's life in _packages.
         """
+        custom_mapping_from_file: Dict[str, List[str]] = {}
 
-        logger.debug(f"Loading user-defined mapping from {self.mapping_path}")
-        with open(self.mapping_path, "rb") as mapping_file:
-            user_mapping = tomllib.load(mapping_file)
+        if self.mapping_path is not None:
+            logger.debug(f"Loading user-defined mapping from {self.mapping_path}")
+            with open(self.mapping_path, "rb") as mapping_file:
+                custom_mapping_from_file = tomllib.load(mapping_file)
 
-        return {
+        mapping = {
             Package.normalize_name(name): Package(
                 name,
                 {DependenciesMapping.USER_DEFINED: set(imports)},
             )
-            for name, imports in user_mapping.items()
+            for name, imports in custom_mapping_from_file.items()
         }
 
+        if self.custom_mapping is not None:
+            logger.debug("Loading user-defined mapping from custom mapping.")
+
+            for name, imports in self.custom_mapping.items():
+                normalised_name = Package.normalize_name(name)
+                if normalised_name in mapping:
+                    logger.info(
+                        "Mapping for %s already found in %s. Import names "
+                        "from the configuration file are appended to ones "
+                        "found in the mapping file.",
+                        normalised_name,
+                        self.mapping_path,
+                    )
+                    mapping[normalised_name].add_import_names(
+                        *imports, mapping=DependenciesMapping.USER_DEFINED
+                    )
+                else:
+                    mapping[normalised_name] = Package(
+                        name,
+                        {DependenciesMapping.USER_DEFINED: set(imports)},
+                    )
+
+        return mapping
```

Comment:
I wonder if the code here becomes simpler if we merge the two dicts _before_ we instantiate the `Package` objects, something like:
```python
from collections import defaultdict

...

        mapping = defaultdict(set)
        for name, imports in custom_mapping_from_file.items():
            normalised_name = Package.normalize_name(name)
            mapping[normalised_name].update(imports)
        for name, imports in self.custom_mapping.items():
            normalised_name = Package.normalize_name(name)
            if normalised_name in mapping:
                logger.info(
                    "Mapping for %s already found in %s. Import names "
                    "from the configuration file are appended to ones "
                    "found in the mapping file.",
                    normalised_name,
                    self.mapping_path,
                )
            mapping[normalised_name].update(imports)

        return {
            name: Package(name, {DependenciesMapping.USER_DEFINED: set(imports)})
            for name, imports in mapping.items()
        }
```

(There are some details around normalization of package names that I might have glossed over here...)

Reply:
This is simpler indeed, but there will be a problem if mapping from settings has a different unnormalised name than mapping from the file (like `SciKit-Learn` vs `scikit-learn`).

The latter will update only with the latest seen mapping in your solution. In the current one, the mappings are merged (see Nour's comment: https://github.com/tweag/FawltyDeps/pull/292#discussion_r1157718469, although the `name` field in `Package` object will be the one first encountered.

Reply:
AFAICS, after another look, all names are normalized before they are put into the `mapping` dict, so imports that are associated to the same _normalized_ name should all end up in the same `Package` object.

The difference is that the `.name` members of the resulting `Package` objects are _also_ normalized in my suggestion, whereas they are not in your diff. I'm not sure that makes any actual difference in practice.

Anyway, this is beyond nitpicking, and the PR is correct AFAICS in any case. Please consider this resolved.